src/fixit/lsp.py

```diff
@@ -0,0 +1,162 @@
+        self.lsp = LanguageServer("fixit-lsp", __version__)
+
+        self._config_cache: Dict[Path, Config] = {}
+        self._debounced_validators: Dict[str, Callable[[int], None]] = {}
+
+        self.set_handlers()
+
+    def load_config(self, path: Path, bust_cache=False) -> Config:
+        """
+        Cached fetch of fixit.toml(s) for fixit_bytes.
+        """
+        if bust_cache or (path not in self._config_cache):
+            self._config_cache[path] = generate_config(path, options=self.fixit_options)
+        return self._config_cache[path]
+
+    def diagnostic_generator(self, uri: str, autofix=False):
+        """
+        LSP wrapper (provides document state from `pygls`) for `fixit_bytes`.
+        """
+        path = Uri.to_fs_path(uri)
+        if not path:
+            return None
+        path = Path(path)
+
+        return fixit_bytes(
+            path,
+            self.lsp.workspace.get_document(uri).source.encode(),
+            autofix=autofix,
+            config=self.load_config(path),
+        )
+
+    def validate(self, uri: str, version: int) -> None:
+        """
+        Side-effect: publishes Fixit diagnostics to the LSP client.
+        """
+        generator = self.diagnostic_generator(uri)
+        if not generator:
+            return
+        diagnostics = []
+        for result in generator:
+            violation = result.violation
+            if not violation:
+                continue
+            diagnostic = Diagnostic(
+                Range(
+                    Position(  # LSP is 0-indexed; fixit line numbers are 1-indexed
+                        violation.range.start.line - 1, violation.range.start.column
+                    ),
+                    Position(violation.range.end.line - 1, violation.range.end.column),
+                ),
+                violation.message,
+                severity=DiagnosticSeverity.Warning,
+                code=violation.rule_name,
+                source="fixit",
+            )
+            diagnostics.append(diagnostic)
+        self.lsp.publish_diagnostics(uri, diagnostics, version=version)
+
+    def debounced_validator(self, uri: str) -> Callable[[int], None]:
+        """
+        Per-URI debounced validation function. See: LSP.validate
+        """
+        if uri not in self._debounced_validators:
+            self._debounced_validators[uri] = debounce(
+                self.lsp_options.debounce_interval
+            )(lambda version: self.validate(uri, version))
+        return self._debounced_validators[uri]
+
+    def set_handlers(self) -> None:
+        """
+        Side-effect: mutates self.lsp to make it respond to
+        `textDocument/didOpen`, `textDocument/didChange`, `textDocument/formatting`
+        """
+
+        @self.lsp.feature(TEXT_DOCUMENT_DID_OPEN)
+        def _(params: DidOpenTextDocumentParams):
+            self.debounced_validator(params.text_document.uri)(
+                params.text_document.version
+            )
```

Comment:
I'm not super fond of the "call a method to get a cached callable and then call that" pattern.  Would prefer if this could be just `self.validate(uri, version)` and let the `validate` method handle debouncing as necessary.